## Goal:

- ### Using GenCompass on WGS BAM files as input

## Data: 

- WGS BAMs

## Code:

- main github from CGR :https://github.com/NCI-CGR/GenCompass_WGS

- since we start from BAM files, need to create some files for input

  - dummpy `fastq_files.txt`,just random for script to not break

 
    ```
    /path/to/fastq/SAMPLE_RUN-ID-6292_S2_L001_R1_001.fastq.gz
    /path/to/fastq/SAMPLE_RUN-ID-6292_S2_L001_R2_001.fastq.gz
    ```

  - `samples.txt` id of list samples
     
  ```
  TCGA_1
  TCGA_2
  ```

  - `mf.csv`, a manifest csv file, the info of sample id, flowcell run, as it start from BAM, flowcell id can input dummy
  - the csv file include header `Sample ID`,`Flowcell`
    
Sample ID|Flowcell|
---------|--------|
TCGA_1|B00FLOWCELL|
TCGA_2|B00FLOWCELL|

  - `set1.json`, copy and renamed from `batch_builder/templates/project_parameters.json`, edit to the file indicated above, also need to download referecne bundle.
    
  - the ref bundle need: hg38 genome sequence fasta. `Homo_sapiens_assembly38_masked_GRC_exclusions.fasta.tar` , and bed file of `wgs_calling_regions.hg38.interval_list.masked_GRC_exclusions.bed`
    
  - the bed file is generated by remove "@" row at the begining of file using `grep -v '^@' wgs_calling_regions.hg38.interval_list > wgs_calling_regions.hg38.interval_list.masked_GRC_exclusions.bed`
    
  - downloaded from GATK reference bundle, still need to figure out remove GCR region

 ```json
{
    "project": "gencompass_workflow_test",
    "manifest": "mf.csv",
    "fastq_files" : "fastq_files.txt",
    "reference_bundle": "./Ref_bd",
    "samples" : "samples.txt",
    "results_directory": "res_test",
    "logs_directory" :"ttlogs",
    "working_directory" : "./",
    "cromwell_jar": "$CROMWELL_JAR",
    "cromwell_config": "$CROMWELL_CONFIG",
    "singularity_cachedir": "/data/$USER/singularity_cache"
}
```
Now for variant calling in this example, specific this function:

```
 python GenCompass_WGS/batch_builder/prepare_gencompass.py variant_calling\
  --project_parameters set1.json\ # <--- the json file we edit previously
  --input_template GenCompass_WGS/batch_builder/templates/variant_calling.input_template.json 

```

it will generate the swarm file `gencompass_workflow_test_variant_calling.swarm`

or just output all of it 


```
python GenCompass_WGS/batch_builder/prepare_gencompass.py \
all \
--project_parameters ./set1.json
```

Once all files are prepared. put the BAM files in the created folder according to generated json `./inputs/variant_calling_inputs/TCGA_1_variant_calling_input.json` as example

the file look like: 

```json
{
    "VariantCalling.sampleID": "TCGA_1",
    "VariantCalling.sampleBQSR": "res_test/fq2bam/TCGA_1/TCGA_1.BQSR-REPORT.txt",
    "VariantCalling.sampleBAM": "res_test/fq2bam/TCGA_1/TCGA_1.bam",
    "VariantCalling.sampleBAI": "res_test/fq2bam/TCGA_1/TCGA_1.bam.bai",
    "VariantCalling.haplotypecallerAdditionalArgs": [],
    "VariantCalling.deepVariantAdditionalArgs": [],
    "VariantCalling.runHaplotypeCaller": true,
    "VariantCalling.runDeepVariant": true,
    "VariantCalling.runStrelka2": true,
    "VariantCalling.gencompassDocker": "cgrlab/gencompass:v1.2",
    "VariantCalling.referenceTarball": "./Ref_bd/Homo_sapiens_assembly38/Homo_sapiens_assembly38_masked_GRC_exclusions.fasta.tar",
    "VariantCalling.strelkaDocker": "cgrlab/strelka:v2.9.10",
    "VariantCalling.callRegions": "./Ref_bd/Homo_sapiens_assembly38/wgs_calling_regions.hg38.interval_list.masked_GRC_exclusions.bed",
    "VariantCalling.parabricksDocker": "cgrlab/clara-parabricks:4.3.0-1",
    "VariantCalling.deepvariantRuntimeAttributes": {
        "memoryGiB": 64,
        "cpuCount": 32,
        "acceleratorType": "v100x",
        "acceleratorCount": 1
    },
    "VariantCalling.haplotypecallerRuntimeAttributes": {
        "memoryGiB": 64,
        "cpuCount": 32,
        "acceleratorType": "v100x",
        "acceleratorCount": 1
    },
    "VariantCalling.strelkaRuntimeAttributes": {
        "memoryGiB": 64,
        "cpuCount": 32
    }

```

Once done, run the swarm file. 

### Next step: 

- Probably need to create the BQSR file?
- have to run in Boiwulf
https://hpc.nih.gov/training/gatk_tutorial/bqsr.html



